pkg/controllers/disruption/drift.go

```diff
@@ -82,16 +82,35 @@ func (d *Drift) ComputeCommand(ctx context.Context, candidates ...*Candidate) (C
 		consolidationTypeLabel: d.ConsolidationType(),
 	}).Set(float64(len(candidates)))
 
+	// Do a quick check through the candidates to see if they're empty.
+	// For each candidate that is empty with a nodePool allowing its disruption
+	// add it to the existing command.
+	empty := make([]*Candidate, 0, len(candidates))
+	for _, candidate := range candidates {
+		if len(candidate.pods) > 0 {
+			continue
+		}
+		// If there's disruptions allowed for the candidate's nodepool,
+		// add it to the list of candidates, and decrement the budget.
+		if disruptionBudgetMapping[candidate.nodePool.Name] > 0 {
+			empty = append(empty, candidate)
+			disruptionBudgetMapping[candidate.nodePool.Name]--
+		}
+	}
 	// Disrupt all empty drifted candidates, as they require no scheduling simulations.
-	if empty := lo.Filter(candidates, func(c *Candidate, _ int) bool {
-		return len(c.pods) == 0
-	}); len(empty) > 0 {
+	if len(empty) > 0 {
 		return Command{
 			candidates: empty,
 		}, nil
 	}
 
 	for _, candidate := range candidates {
+		// If the disruption budget doesn't allow this candidate to be disrupted,
+		// continue to the next candidate. We don't need to decrement any budget
+		// counter since drift commands can only have one candidate.
+		if disruptionBudgetMapping[candidate.nodePool.Name] == 0 {
```

Comment:
Is it possible to emit a debug log or metric around the number of candidates that were "left behind" because budget has been drained.

Reply:
Definitely will emit a metric as a followup. I worry about a log here, since this may get very noisy, and we'd need to duplicate this across each method, and it would probably need to log per-provisioner, per-method, per-disruption loop, which I could envision getting noisy. @jonathan-innis and I were talking about modeling this as a status condition per-node. Thoughts on this - metric for followup, and status condition for node - rather than emitting a log?

Reply:
Ya that sounds good, I think I just want a way to know if we are falling behind and by how much in addition to which are the nodes falling behind